### DISCUSSION TITLE
Pseudocode

### DISCUSSION THREAD

**Discussion by jdlocklin** [ftc-saturn9944/FtcRobotController#2] [General]
Just some preliminary pseudocode thoughts:

The two most important functions will be detecting our cone and lifting. The following thoughts address possible solutions for accomplishing these tasks.

I think it's essential to have a getLiftDistance() function (reads current height in inches of distance sensor mounted on lift) and getColorDistance() function (reads same thing for distance sensor on v-plate)

Then we should also design a built-in grip() and lift( height double) function on our init class.

public static void grip() {
       robot.gripper.setPosition(GRIP);
}

public static void lift( height double) { 
     while (getLiftDistance < height) {
          robot.lift.setPower(LIFTING POWER);
     }
     robot.lift.setPower(STATIC LIFTING POWER); or just brake it.
}

then for actual gamepad input for lifting:
 if(button pressed)
get distance to lift to() 
will get current lift height and use a for loop to iterate through the array of set heights and return the next step up.
Then pass it'll that height as an argument for lift().

I also think the position of the gripper should define whether we are intaking or scoring.
so
public static boolean intakeToggle() {
      if(robot.gripper.getPosition() == GRIP) {
           return false;
      } else {
            return true

So, if the v-plate color sensor detects its distance is less than the set distance we have for it and intakeToggle() is true, then it grips a cone, regardless of the height the lift is raised to. 
If it's false then that means we are at a junction.

**Comment by jleechpe** [comment]
Using FTC Lib the code for the lift can leverage the `.atTarget()` method to know if it's at the desired height or not.  This uses a target distance + threshold (+/- X) to determine if it has reached the target height.  Can use this same atTarget/below/above logic for raising/lowering via buttons (set currentTarget then `liftMotor.moveTo(currentTarget)`) and during pivot (by setting a current target height that is high enough to pivot safely.

```java
if (liftSensor.atTarget(currentTarget)) {
  liftMotor.setSpeed(0);
} else if (liftSensor.getDistance() < currentTarget.getTarget()) {
  liftMotor.setSpeed(+0.3);
} else {
  liftMotor.setSpeed(-0.3);
}
```

Then we need 2 sets of [DistanceTargets](https://github.com/FTCLib/FTCLib/blob/v2.0.1/core/src/main/java/com/arcrobotics/ftclib/hardware/SensorDistanceEx.java#L15) + 1 fixed distance:
- Fixed : Minimum height to rotate to other side.  Use this in the "flip sides" logic to lift to that height if below it before rotating.
    ```java
      Boolean tooLow = liftSensor.getDistance() < pivotHeight
      if (tooLow) {
        var currHeight = liftSensor.getDistance() // should be a variable for both uses just here
        liftMotor.moveTo(pivotHeight);
      }
      // do pivot here
      if (tooLow) {
        liftMotor.moveTo(currHeight);
      }
    ```
- Set1 : Intake heights (Floor, 2, 3, 4, 5 cones)
- Set2 : Scoring heights (Junction, Low, Mid, High)

Then we either use `intake` or `scoring` LiftHeights depending on whether we have a cone or not.

A custom class for the LiftHeight targets.  This allows `.next()` and `.previous()` to step up and down the list of targets (using gamepad buttons/triggers for "Lift" (`.next()`) and "Lower" (`.previous()`)).  If the lift somehow ends up between those heights it will move to the next target height on button press.  The button would update the target height then call the `moveTo(<target>)` method.

```java
class LiftHeights {
  private List<DistanceTarget> heights;
  private DistanceTarget currentTarget;

  /* Missing code
  // Constructors to create the list of heights
  // Way to add a list of targets or an individual target
  */

  public DistanceTarget getTarget {
    return currentTarget;
  }

  // Needs logic to fail gracefully if name doesn't exist
  public void setTarget (String name) {
    currentTarget = heights.stream().filter(x -> x.name == name)
  }

  public void next() {
    /*
    // Get the index of currentTarget from the list of Targets
    // Initialize ListIterator at Index+1 (ListIterator is between 2 indices so index 1 will return 1 for `.next()`)
    */
    if (iterator.hasNext()) {
      currentTarget = iterator.next()
    }
  }

  public void previous() {
    /*
    // Get the index of currentTarget from the list of Targets
    // Initialize ListIterator at Index
    */
    if (iterator.hasPrevious()) {
      currentTarget = iterator.previous()
    }
  }
```

**Comment by jleechpe** [comment]
The reason for the `.hasNext()` and `.hasPrevious()` logic above is that it will automatically prevent the value from cycling higher/lower than the last/first target, meaning you won't go further up if at the top (or lower down at the bottom).  Since an ArrayList is ordered this also means stepping up/down the list will always be done in order.

DistanceTarget takes `(DistanceUnit unit, Double targetDistance, double threshold = 5, string name = "Distance Target")` (I'm putting the defaults inline as `<T var = defaultValue>`).  Using the name property means setting a specific target (for Autonomous or say when swapping between intake and scoring and wanting to move to the preferred height right away) just requires the name (and makes the code easier to read when looking it over).
